java-connectors/kafka-connect-azure-eventhubs/src/main/java/io/lenses/java/streamreactor/connect/azure/eventhubs/source/EventHubsKafkaConsumerController.java

```diff
@@ -0,0 +1,144 @@
+package io.lenses.java.streamreactor.connect.azure.eventhubs.source;
+
+import static io.lenses.java.streamreactor.connect.azure.eventhubs.config.AzureEventHubsConfig.getPrefixedKafkaConsumerConfigKey;
+import static io.lenses.java.streamreactor.connect.azure.eventhubs.mapping.SourceRecordMapper.mapSourceRecordIncludingHeaders;
+import static io.lenses.java.streamreactor.connect.azure.eventhubs.mapping.SourceRecordMapper.mapSourceRecordWithoutHeaders;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.GROUP_ID_CONFIG;
+
+import io.lenses.java.streamreactor.connect.azure.eventhubs.config.AzureEventHubsConfig;
+import io.lenses.java.streamreactor.connect.azure.eventhubs.config.AzureEventHubsConfigConstants;
+import io.lenses.java.streamreactor.connect.azure.eventhubs.source.TopicPartitionOffsetProvider.AzureOffsetMarker;
+import io.lenses.java.streamreactor.connect.azure.eventhubs.source.TopicPartitionOffsetProvider.AzureTopicPartitionKey;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.config.SaslConfigs;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.source.SourceRecord;
+
+/**
+ * Class is a bridge between EventHub KafkaConsumers and AzureEventHubsSourceTask. It verifies the configuration
+ * of kafka consumers and instantiates them, then allows AzureEventHubsSourceTask to pull for SourceRecords.
+ */
+@Slf4j
+public class EventHubsKafkaConsumerController {
+
+  private final AzureEventHubsConfig azureEventHubsConfig;
+  private final KafkaConsumerProvider kafkaConsumerProvider;
+  private final BlockingQueue<ConsumerRecords<String, String>> recordsQueue;
+  private final BlockingQueuedKafkaConsumer queuedKafkaConsumer;
```

Comment:
This class has two purposes of now:
- It's a consumer of BlockingQueue whenever Task calls `poll()`
- It coordinates Kafka `Consumer` initialization by calling `KafkaConsumerProvider`

That's why it has both of those.

Reply:
A Task configuration is fixed once the Task is created. I see the logic to deal with configuration present in this class but also in KafkaConsumerProvider:

```
  KafkaConsumer<String, String> kafkaConsumer = new KafkaConsumer<>(consumerProperties);
    String topic = consumerProperties.getProperty(AzureEventHubsConfigConstants.EVENTHUB_NAME);
    return new BlockingQueuedKafkaProducer(recordBlockingQueue, kafkaConsumer, clientId, topic);
```

I think it would be better to separate and consolidate creating a Kafka consumer from the task properties, including the group and instance id. Doing so allows you to test this code in isolation, and you are not coupling the controller with the configuration either.

The KafkaConsumerProvider should give back an instance of a Consumer<Bytes, Bytes> and take care of the underlying. And in the interest of testing would it not make sense for KafkaConsumerProvider to be an interface?

Reply:
Agreed on removing code that dealt with properties from here and delegated to `ProducerProvider`.
This also means we don't have to keep some of those fields and class if more in-line with _Single Responsibility Principle_.

About interface: I added it as it will pay off when we abstract from String (and if we ever abstract from Kafka to let's say Azure SDK)